src/Runner.Worker/Container/ContainerHooks/ContainerHookManager.cs

```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using GitHub.DistributedTask.Pipelines.ContextData;
+using GitHub.DistributedTask.WebApi;
+using GitHub.Runner.Common;
+using GitHub.Runner.Common.Util;
+using GitHub.Runner.Sdk;
+using GitHub.Runner.Worker.Handlers;
+using GitHub.Services.WebApi;
+
+namespace GitHub.Runner.Worker.Container.ContainerHooks
+{
+    [ServiceLocator(Default = typeof(ContainerHookManager))]
+    public interface IContainerHookManager : IRunnerService
+    {
+        Task PrepareJobAsync(IExecutionContext context, List<ContainerInfo> containers);
+        Task CleanupJobAsync(IExecutionContext context, List<ContainerInfo> containers);
+        Task ContainerStepAsync(IExecutionContext context);
+        Task RunScriptStepAsync(IExecutionContext context);
+    }
+
+    public class ContainerHookManager : RunnerService, IContainerHookManager
+    {
+        private const string ResponseFolderName = "_hook_responses";
+        private string HookIndexPath;
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            HookIndexPath = $"{Environment.GetEnvironmentVariable(Constants.Hooks.ContainerHooksPath)}";
+        }
+        
+        public async Task PrepareJobAsync(IExecutionContext context, List<ContainerInfo> containers)
+        {
+            Trace.Entering();
+            var responsePath = GenerateResponsePath();
+            var jobContainer = containers.Where(c => c.IsJobContainer).FirstOrDefault();
+            var serviceContainers = containers.Where(c => c.IsJobContainer == false).ToList();
+
+            var input = new HookInput
+            {
+                Command = HookCommand.PrepareJob,
+                ResponseFile = responsePath,
+                Args = new HookArgs
+                {
+                    JobContainer = jobContainer.GetHookContainer(),
+                    Services = serviceContainers.Select(c => c.GetHookContainer()).ToList(),
+                }
+            };            
+
+            var response = await ExecuteHookScript(context, input);
+            context.JobContext.Container["id"] = new StringContextData(response.Context.Container.Id);
```

Comment:
These all nullref is no context is applied, we should wrap them in an if statement, I'm thinking context is an optional return type.

If we end up making it required, we could just add a better error message on the if

Reply:
Another interesting tibet here, if we dont' set these values other things start failing, like `ActionRunner` line 162

```
ArgUtil.NotNullOrEmpty(ExecutionContext.Global.Container.ContainerId, nameof(ExecutionContext.Global.Container.ContainerId));
```

We will need to audit everywhere that expects them to be set, and not require it if using hooks

Reply:
> These all nullref is no context is applied

Like if there was no context written in the response file? Agreed. If this context doesn't exist, it undermines the rest of _our_ docker hooks, like run_script_step, but does it make sense to enforce it for all implementations of prepare_job hooks? (Or just all docker hooks for now?)

Reply:
> If this context doesn't exist, it undermines the rest of our docker hooks, like run_script_step

I think currently this is the case in the runner, but I think we want to refactor this, I don't think it should.

From a purity of purpose standpoint, the hook is 

- `create the containers for the job`
- `run a step in the job container`
The hook is responsible for keeping track of the job container id (if it even has one pods do not), so it can exec whatever step we pass into it next.

The context is used for expression evaluation by users, if the hooks don't return those values, the expression evaluation should just be empty, it shouldn't stop/break the users build.

For example, if in k8s we don't set `job.container.network`. That should be fine, we don't have a network in k8s, it should either be networked and just work, or we won't support those networking features.

Reply:
We can add a todo and come back to this in another pr if thats easier!

Reply:
I'll do that, let's get the stepscript moving!